### PR TITLE
Bar hit detection in simulation

### LINKED ISSUE #41: Gazebo simulation of contact with keys
Gazebo simulation to simulate contact with keys & get which note was played. Requires changes to the URDF model (fixed -> prismatic joints for bars?)

### LINKED ISSUE #104: Add documentation for simulation
Add docs to `README.md` and to sim package with more details on how everything works & connects to each other.

### LINKED ISSUE #163: Generate MIDI from notes detected in simulation

### PR DESCRIPTION
**Pull request by iibrahimli**
Detect bar contact based on joint velocity
Adds a node `detect_bar_contact` and a launchfile to run RViz, Gazebo, detect_bar_contact, and plotjuggler:
```
roslaunch marimbabot_simulation marimbabot_contact_debug.launch
```
The node `detect_bar_contact` detects which bars (notes) were hit based on `joint_states` and publishes a note string (e.g. `b5`, `dis5`) to topic `sim_notes`. The node `play_notes` then plays the detected notes using a crude (for now) approach with real WAV files recorded in a very noisy room.

A MIDI file is generated and saved to `marimbabot_simulation/midi` when the nodes are stopped.

Closes #163
Closes #41
Closes #104

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -20,4 +20,5 @@ transformers[sentencepiece]==4.26.0
 catkin_pkg==0.5.2
 openai-whisper==20230314
 webrtcvad==2.0.10
-wave==0.0.2
\ No newline at end of file
+wave==0.0.2
```
Comment (Juphex) [inline requirements.txt:23]: should be installed by default https://docs.python.org/3/library/wave.html

File: marimbabot_simulation/src/play_notes.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  "gis5",
+  "gis6",
+]
+
+
+def find_file_directory(filename):
+    home_directory = os.path.expanduser("~")
+    command = f"find {home_directory} -type f -name {filename} -exec dirname {{}} \\; -quit 2>/dev/null"
+    try:
+        output = subprocess.check_output(command, shell=True, universal_newlines=True)
```
Comment (Juphex) [inline marimbabot_simulation/src/play_notes.py:58]: why don't you use Python for the whole code (os.walk etc.) or this line of code?
result = os.popen(command).read().strip()

Just out of interest

Comment (iibrahimli) [reply marimbabot_simulation/src/play_notes.py:58]: Habit. It's also faster (probably)

**Review by berkgungor — APPROVED** [review]
overall LGTM but i am not sure if it makes sense to keep all the audio files in the repo. There are so many wav files.

**Review by Juphex — CHANGES_REQUESTED** [review]
I cannot see the launch file you mention

"marimbabot_contact_debug.launch"

@iibrahimli

**Comment by iibrahimli** [comment]
> I cannot see the launch file you mention
> 
> 
> 
> "marimbabot_contact_debug.launch"
> 
> 
> 
> @iibrahimli 
> 
> 

PR description is outdated, should be marimbabot_simulation/launch/marimbabot_full_sim.launch as shown in the Readme

**Comment by Juphex** [comment]
@iibrahimli 
I had to add
sudo apt-get install libportaudio2  (additionally to the python package)
and 
[chomp_motion_planner](http://wiki.ros.org/action/fullsearch/chomp_motion_planner?action=fullsearch&context=180&value=linkto%3A%22chomp_motion_planner%22)
in the package.xml
and install ros-noetic-moveit-planners-chomp

Please consider adding them :)

**Comment by iibrahimli** [comment]
> @iibrahimli I had to add sudo apt-get install libportaudio2 (additionally to the python package) and [chomp_motion_planner](http://wiki.ros.org/action/fullsearch/chomp_motion_planner?action=fullsearch&context=180&value=linkto%3A%22chomp_motion_planner%22) in the package.xml and install ros-noetic-moveit-planners-chomp
> 
> Please consider adding them :)

Good catch! thanks, will add now.

**Comment by iibrahimli** [comment]
> overall LGTM but i am not sure if it makes sense to keep all the audio files in the repo. There are so many wav files.

Agreed. Now that you mention it, I will remove the notes for now, we generate a MIDI file and can play that anyways.

**Comment by iibrahimli** [comment]
> 

Regarding chomp stuff, no motion planning is done in the simulation package. I'd guess that is a dependency from another package.

File: marimbabot_simulation/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+* `marimbabot_full_sim.launch`: Launches the simulation in Gazebo, as well as nodes to detect contacts and publish the contact points, and generate MIDI.
+
+
+## How to run
+
+Install dependencies (if not already installed):
+
+```bash
+sudo apt-get install libportaudio2
+sudo apt-get install fluidsynth
```
Comment (Juphex) [inline marimbabot_simulation/README.md:17]: you can remove fluidsynth, it is already installed here https://github.com/UHHRobotics22-23/marimbabot/blob/6341875dd5ed0e60db247a12c4871d0d73637701/marimbabot_audio/package.xml#L19

**Review by Juphex — APPROVED** [review]
LGTM, as we talked about it @iibrahimli

**Pull request merged**